### PR TITLE
Pass Params and Search instead of context

### PR DESCRIPTION
**Pull request by NicholasHallman**
I think this is the last place we were using the page context. This should introduce enough isolation that if we ever need to we can abandon page.js

Quality Notes

### Testing
 - [x] Test parameters are available in the view
 - [x] Test search params are available in the view 
 - [x] Test both are available

### Documentation

Updated the readme to demonstrate param and search passing to the view function.

FYI: @anhill-D2L @johngwilkinson  @bemailloux

### CODE REVIEW

File: example/people/route-loader.js [hunk 1]
```diff
@@ -7,19 +7,17 @@ export function loader() {
         {
             pattern: '/people',
             loader: () => import('./people.js'),
-            view: context => html` <test-people
+            view: (_, search) => html` <test-people
```
Comment (dlockhart) [inline example/people/route-loader.js:10]: I do like the named parameters, but I am worried that in the future we'll need more parameters and if you just need search you'll need to go `(p1, p2, p3, p4, search) => ` instead of just `(ctx) =>`. 🤔

File: src/router.js [hunk 2]
```diff
@@ -6,8 +6,12 @@ let _lastOptions = {};
 
 const _handleRouteView = (context, next, r) => {
     if (r.view) {
-        context.view = r.view(context);
+        context.view = r.view.apply(null, [
+            context.params,
+            context.searchParams,
+        ]);
```
Comment (dlockhart) [inline src/router.js:12]: I remember [reading here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions) that arrow functions aren't suitable for `apply` since there's no scope. Would `r.view(context.params, context.searchParams)` work?

File: README.md [hunk 3]
```diff
@@ -12,18 +12,56 @@ registerRoutes([
...
+```js 
+pattern: '/user/:id/:page'
+view: (params) => html`<user-view id=${params.id} page=${params.page}></user-view>` 
+```
+
+A `URLSearchParameters` object is always the last parameter of the view function:
+
+```js 
+pattern: '/user' // search: ?id=1&page=1&semester=1
+view: (_, search) => html`
```
[the remark this answers is not in the document]
Comment (NicholasHallman) [reply README.md]: I like that idea, then we can include path as well. Although it's somewhat implied since you are on that route lol

File: README.md [hunk 4]
```diff
@@ -52,7 +90,7 @@ class FooBar extends LitElement {
 
     render() {
         let userId = this.route.params.userId;
-        let orgId = this.route.search.get('org-unit');
+        let orgId = this.route.search['org-unit'];
```
[the remark this answers is not in the document]
Comment (NicholasHallman) [reply README.md:88]: Here's another time I wish js had some rust features. Rust's result pattern would work great in this example.

`search.get('org-unit').unwrap_or(undefined)` <- undefined is now returned if the value is null

File: example/people/route-loader.js (hunk 1 above)
Comment (NicholasHallman) [reply example/people/route-loader.js:10]: This PR doesn't do named parameters it groups all the parameters together as the first input, and the search as the second one so

`/user/:id/:orgUnit/:semester : (params, search) => html`params.id params.orgUnit params.semester` `

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (dlockhart) [reply README.md:88]: Maybe we create our own `ifNotNullAndDefined` directive haha... only half joking I think.

File: example/people/route-loader.js (hunk 1 above)
Comment (dlockhart) [reply example/people/route-loader.js:10]: Yeah sorry this comment was a repeat of the earlier one ("If we foresee any more parameters in the future this pattern...") that I thought got lost during my review when I reloaded the page. Ignore.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (johngwilkinson) [reply README.md:88]: `const orgId = this.route.search['org-unit'] || undefined`?

[the remark this answers is not in the document]
Comment (dlockhart) [reply README.md:88]: Yep that would totally work, just doesn't have the same ring to it as `ifDefined(this.route.search.get('orgUnit'))`. :)

File: README.md [hunk 5]
```diff
@@ -58,6 +58,62 @@ view: (params, search) => html`
     </user-view>` 
 ```
 
+### Multiple Route Loaders
+
+It is encouraged that you use a src folder structure like this.
```
Comment (dlockhart) [inline README.md:63]: I think it's less important that each "page" load its own routes as that seems like overkill. This would more be for situations where there were significant "sub-applications" within a larger application. So I'd rephrase this documentation to target more those larger applications vs. "it is encourages that all apps do this".

**Review by dlockhart — APPROVED** [review]
